### PR TITLE
Expose translog stats in ReadOnlyEngine

### PR DESCRIPTION
**Pull request by tlrx**
This pull request changes `ReadOnlyEngine` so that it loads stats from the translog when the engine is created. It also adds tests to check that translog stats are correctly exposed in Stats API for closed and frozen indices.

### CODE REVIEW

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/docbldesx

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml [hunk 1]
```diff
@@ -77,3 +77,51 @@ setup:
       indices.stats:
         metric: [ translog ]
   - gte: { indices.test.primaries.translog.earliest_last_modified_age: 0 }
+
+---
+"Translog stats on closed indices":
+  - skip:
+      version: " - 7.2.99"
```
Comment (dnhatn) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml]: I think we should skip this test until 7.9.99 then adjust to this value after backporting.

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 2]
```diff
@@ -119,6 +121,7 @@ public ReadOnlyEngine(EngineConfig config, SeqNoStats seqNoStats, TranslogStats
                 reader = wrapReader(reader, readerWrapperFunction);
                 searcherManager = new SearcherManager(reader, searcherFactory);
                 this.docsStats = docsStats(lastCommittedSegmentInfos);
+                this.translogStats = translogStats != null ? translogStats : translogStats(config, lastCommittedSegmentInfos);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: Can we assert that if `translogStats` is null then `obtainLock` is true? This ensures that we don't open two translog instances at the same time.

**Review by dnhatn — APPROVED** [review]
LGTM. I left two minor comments.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml (hunk 1 above)
Comment (ywelsch) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml]: weird that this did not fail the tests? Perhaps BWC is disabled?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml [hunk 3]
```diff
@@ -77,3 +77,51 @@ setup:
...
+
+  - do:
+      indices.close:
+        index: test
+  - is_true: acknowledged
+
+  - do:
+      cluster.health:
+        expand_wildcards: all
+        wait_for_no_initializing_shards: true
```
Comment (ywelsch) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml]: this  is only needed for the backport.
Alternatively you could use wait_for_active_shards = 1 on the indices.close action

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 4]
```diff
@@ -80,6 +82,7 @@
     private final DocsStats docsStats;
     private final RamAccountingSearcherFactory searcherFactory;
 
+
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: undo this

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java (hunk 2 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: ++ especially as opening the translog here writes a new checkpoint file.

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml (hunk 1 above)
Comment (tlrx) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml]: Pfff... I should have seen this. And yes, bwc tests were disabled at that time. I merged master and bwc are enabled again.

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java (hunk 2 above)
Comment (tlrx) [reply server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: This is a very good suggestion, thanks

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml (hunk 3 above)
Comment (tlrx) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/20_translog.yml]: Right

**Comment by tlrx** [comment]
@dnhatn @ywelsch Thanks for your feedback. I updated the code.

**Review by ywelsch — APPROVED** [review]
LGTM. Don't forget to add the wait_for_active_shards condition on the 7.x backport.

**Comment by tlrx** [comment]
Thanks @ywelsch and @dnhatn

**Pull request merged**